File: roseWong/assignments/slidingwindow/lc209/lc209.py
# Problem Statement #

# Given an array of positive numbers and a positive number ‘S’,
# find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’.
# Return 0 if no such subarray exists.

# starting with the code from 

import math


def smallest_subarray_with_given_sum(s, arr):
  minLength = math.inf
  windowLength = 0
  windowSum = 0
  windowStart = 0
  windowEnd = 0
  for windowEnd in range(len(arr)):
    windowSum = windowSum + arr[windowEnd]  # this adds the next element in the array
    windowLength = windowLength + 1
    # slide the window when we hit the required size of s
    while windowSum >= s:
      minLength = min(minLength, (windowEnd - windowStart + 1))
      # remove the element that is being dropped from the next sum
      windowSum = windowSum - arr[windowStart]
      windowStart = windowStart + 1
      if minLength == math.inf:
        return 0
  return minLength


def main():
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(7, [2, 1, 5, 2, 3, 2])))
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(7, [2, 1, 5, 2, 8])))
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(8, [3, 4, 1, 1, 6])))


main()



  # Solution
  # -----
  # window_sum = 0
  # min_length = math.inf
  # window_start = 0

  # for window_end in range(0, len(arr)):
  #   window_sum += arr[window_end]  # add the next element
  #   # shrink the window as small as possible until the 'window_sum' is smaller than 's'
  #   while window_sum >= s:
  #     min_length = min(min_length, window_end - window_start + 1)
  #     window_sum -= arr[window_start]
  #     window_start += 1
  # if min_length == math.inf:
  #   return 0
  # return min_length

File: serdaSimus/assignments/slidingwindow/lc209/lc209.js
// Problem Statement #

// Given an array of positive numbers and a positive number ‘S,’ find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’. Return 0 if no such subarray exists.

/* 
in: an array, an integer
out: windowSize >= 0

Solution:
smallestWindowSize
currentWindowSize
windowStart
windowEnd(current index)
sum: as my running total


visit every item, Keep a running sum of the items visited.
    record where the windows started, and its current position.
    windowSize is its current pos less its start pos + 1 because array start at 0.
    each time we add a new value to sum, increase windowSize.
    each time we remove a value from sum, shrink windowSize.
    
*/

function smallest_subarray_with_given_sum(s, arr) {
  // TODO: Write code here
  let smallestWindowSize = Infinity,
    currentWindowSize = 0,
    windowStart = 0,
    sum = 0;

  for (let windowEnd in arr) {
    if (arr[windowEnd] >= s) {
      sum = arr[windowEnd];
      windowStart = windowEnd;
      currentWindowSize = 1;
      smallestWindowSize =
        currentWindowSize < smallestWindowSize
          ? currentWindowSize
          : smallestWindowSize;
      break;
    } else {
      sum += arr[windowEnd];
      currentWindowSize++;

      while (sum >= s) {
        // check if the window size we have is smaller than the previous window size.
        smallestWindowSize =
          currentWindowSize < smallestWindowSize
            ? currentWindowSize
            : smallestWindowSize;

        // substract the value at windowStart
        // decrease windowSize
        // increase windowStart index.
        sum -= arr[windowStart];
        windowStart++;
        currentWindowSize--;
        continue;
      }
    }
  }
  return smallestWindowSize;
}

console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 3, 2]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 8]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    8,
    [3, 4, 1, 1, 6]
  )}`
);

// Solution
// -----
// let windowSum = 0,
// minLength = Infinity,
// windowStart = 0;

// for (windowEnd = 0; windowEnd < arr.length; windowEnd++) {
// windowSum += arr[windowEnd]; // add the next element
// // shrink the window as small as possible until the 'window_sum' is smaller than 's'
// while (windowSum >= s) {
//   minLength = Math.min(minLength, windowEnd - windowStart + 1);
//   windowSum -= arr[windowStart];
//   windowStart += 1;
// }
// }

// if (minLength === Infinity) {
// return 0;
// }
// return minLength;
